lib/engine/apps/exec_wrapper.js

```diff
@@ -129,7 +133,31 @@ class ExecWrapper extends ExecEnvironment {
         return deviceView.values();
     }
 
+    enterProcedure(procid, procname) {
+        // save the calling frame ID on the stack
+        this._procedureStack.push(this._procedureFrame);
+
+        // make a fresh ID for the new call
+        this._procedureFrameCounter++;
+        this._procedureFrame = this._procedureFrameCounter;
+    }
+
+    addExitProcedureHook(hook) {
+        this._hooks.push(hook);
+    }
+    async exitProcedure() {
+        // check that enter & exit are correctly paired
+        assert(this._procedureStack.length > 0);
+
+        // restore the frame ID of the caller from the stack
+        this._procedureFrame = this._procedureStack.pop();
```

Comment:
Same here, we can call `await super.exitProcedure()` and avoid code duplication.